source/common/network/listener_filter_buffer_impl.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/network/io_handle.h"
+#include "envoy/network/listener_filter_buffer.h"
+
+#include "source/common/buffer/buffer_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+using ListenerFilterBufferOnCloseCb = std::function<void()>;
+using ListenerFilterBufferOnDataCb = std::function<void()>;
+
+enum class PeekState {
+  // Peek data status successful.
+  Done,
+  // Need to try again.
+  Again,
+  // Error to peek data.
+  Error
+};
+
+class ListenerFilterBufferImpl : public ListenerFilterBuffer, Logger::Loggable<Logger::Id::filter> {
+public:
+  ListenerFilterBufferImpl(IoHandle& io_handle, Event::Dispatcher& dispatcher,
+                           ListenerFilterBufferOnCloseCb close_cb,
+                           ListenerFilterBufferOnDataCb on_data_cb, uint64_t buffer_size);
+
+  // ListenerFilterBuffer
+  const Buffer::ConstRawSlice rawSlice() const override;
+  // ListenerFilterBuffer
+  bool drain(uint64_t length) override;
+  // ListenerFilterBuffer
+  uint64_t length() const override { return data_size_; }
+
+  /**
+   * Trigger the data peek from the socket.
+   */
+  PeekState peekFromSocket();
+
+  void reset() { io_handle_.resetFileEvents(); }
+
+private:
+  void onFileEvent(uint32_t events);
+
+  IoHandle& io_handle_;
+  Event::Dispatcher& dispatcher_;
+  ListenerFilterBufferOnCloseCb on_close_cb_;
+  ListenerFilterBufferOnDataCb on_data_cb_;
+
+  // The buffer for the data peeked from the socket.
+  std::unique_ptr<uint8_t[]> buffer_;
+  // The start of buffer.
+  uint8_t* base_;
+  // The size of buffer;
+  uint64_t buffer_size_;
+  // The size of valid data.
+  uint64_t data_size_{0};
```

Comment:
Can you remind me why we are not just using a wrapped BufferImpl? It seems strange that we would reimplement all of this logic?

Reply:
The major reason is PEEK always overrides the existing data in the buffer(Instance's reserve/commit is about appending data), and `recv` call expects a linearize buffer. In the end, I reserve a single slice in the constructor, and still accounting the valid data in that single slice. Then we agree that it didn't provide too much value here.

The discussion and the code is still here:
https://github.com/envoyproxy/envoy/pull/17395#discussion_r701288390

Reply:
OK if this was already discussed with @ggreenway I'm fine leaving this as is, however, can we potentially add some basic fuzzing of this class to the buffer fuzz test? That would make me feel a bit more comfortable. We are duplicating some C style memory management here which makes me a bit twitchy from a security perspective.

Reply:
Yeah, I had the same feeling about more low-level buffer management. But using a BufferImpl really didn't fit well and needed more management on top that was redundant. But I'm in favor of some fuzzing.

Reply:
got it, let me add fuzzing test for it.